### PR TITLE
feat(ndt_scan_matcher): remake diag

### PR DESCRIPTION
**Pull request by YamatoAndo**
## Description

remake the ndt_scan_matching's diagnostics.

### before (2023/09/22)
Screenshot from 2023-09-22 18-44-00

---

### after
outputs 6 messages

#### diagnostic for scan_matching_status
diagnostic_scan_matching_status

#### diagnostic for initial_pose_subscriber_status
diagnostic_initial_pose_subscriber_status

#### diagnostic for regularization_pose_subscriber_status
diagnostic_regularization_pose_subscriber_status

#### diagnostic for trigger_node_service_status
diagnostic_trigger_node_service_status

#### diagnostic for ndt_align_service_status
diagnostic_ndt_align_service_status

#### diagnostic for map_update_status
diagnostic_map_update_status

If you want to know the conditions for transitioning to WARN/ERROR, please read the [README.md](https://github.com/autowarefoundation/autoware.universe/blob/fd147ff78de55298cd827d4e57aae5e5e851b958/localization/ndt_scan_matcher/README.md).

## Related links

ToDo
- [x] moidfy system_error_monitor's config file
https://github.com/autowarefoundation/autoware.universe/tree/main/system/system_error_monitor/config
https://github.com/autowarefoundation/autoware_launch/tree/main/autoware_launch/config/system/system_error_monitor
- [x] modify fault_injection.param.yaml
https://github.com/autowarefoundation/autoware.universe/blob/main/simulator/fault_injection/config/fault_injection.param.yaml
https://github.com/autowarefoundation/autoware_launch/blob/main/autoware_launch/config/simulator/fault_injection.param.yaml
- [x] modify system_diagnostic_monitor's config file https://github.com/autowarefoundation/autoware.universe/blob/main/system/system_diagnostic_monitor/config/localization.yaml

merge with
- https://github.com/autowarefoundation/autoware.universe/pull/6889
- https://github.com/autowarefoundation/autoware_launch/pull/973
 

## Tests performed
Lsim works 

## Notes for reviewers

N/A

## Interface changes

message of  `/diagnostics` topic

## Effects on system behavior

N/A

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by YamatoAndo** [comment]
I plan to resume work in December or January.

**Comment by YamatoAndo** [comment]
sorry. I will put it on hold for a while.

File: localization/ndt_scan_matcher/src/map_update_module.cpp
Comment (SakodaShintaro) [inline localization/ndt_scan_matcher/src/map_update_module.cpp:82]: I think it would be better to incorporate `should_update_map` into `update_map` and remove the if statement here.

**Review by SakodaShintaro — COMMENTED** [review]
As a general policy, I felt that it would be better not to separate the validation function in `ndt_scan_matcher`, but to add it where the error occurs in the function being operated.

File: localization/ndt_scan_matcher/CMakeLists.txt
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/CMakeLists.txt:54]: To avoid accidentally adding it, I'm deliberately leaving it commented out.

File: localization/ndt_scan_matcher/CMakeLists.txt
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/CMakeLists.txt:63]: To avoid accidentally adding it, I'm deliberately leaving it commented out.

File: localization/ndt_scan_matcher/src/diagnostics_module.cpp
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/src/diagnostics_module.cpp:68]: I'd like it to be fixed to make it easier to understand what it's doing at a glance.

File: localization/ndt_scan_matcher/src/map_update_module.cpp
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/src/map_update_module.cpp:100]: I thought it would be better to keep the value, but I'll think about it a bit more.

File: localization/ndt_scan_matcher/src/diagnostics_module.cpp
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/src/diagnostics_module.cpp:54]: I've thought about it, but fundamentally, I don't think it will be used.

File: localization/ndt_scan_matcher/src/map_update_module.cpp
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/src/map_update_module.cpp:82]: `update_map` is also used in `ndt_align_service`, so I don't think it can be simply incorporated.

File: localization/ndt_scan_matcher/include/ndt_scan_matcher/map_update_module.hpp
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/include/ndt_scan_matcher/map_update_module.hpp]: fixed https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/51cdedc147536225763b6a6d2782a1c3b07b16d0

File: localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp
Comment (SakodaShintaro) [inline localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:209]: I like this style.
I think it's a good idea to construct all callbacks with three lines like this to create a unified style.

```
diagnostics_xxx_->clear();
main_process();
diagnostics_xxx_->publish();
```

callbacks

* callback_timer() : OK
* callback_initial_pose() : OK
* callback_regularization_pose() : OK
* callback_sensor_points () 
 : The style may be a little different　https://github.com/autowarefoundation/autoware.universe/blob/5f8391846b7922689f7b371b5bc731c746cf16fc/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp#L265-L309
* MapUpdateModule::callback_timer() : The style may be a little different https://github.com/autowarefoundation/autoware.universe/blob/5f8391846b7922689f7b371b5bc731c746cf16fc/localization/ndt_scan_matcher/src/map_update_module.cpp#L67-L74

File: localization/ndt_scan_matcher/src/diagnostics_module.cpp
Comment (SakodaShintaro) [inline localization/ndt_scan_matcher/src/diagnostics_module.cpp:137]: I think the following changes made these codes unnecessary.
https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/f975f1bee7ef607fe1d545859db45a541f43ef4a

Comment (YamatoAndo) [reply localization/ndt_scan_matcher/src/diagnostics_module.cpp]: fixed https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/3cd9471c9952010916a18064ca83561971d156a7

File: localization/ndt_scan_matcher/src/diagnostics_module.cpp
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/src/diagnostics_module.cpp]: fixed https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/f975f1bee7ef607fe1d545859db45a541f43ef4a

File: localization/ndt_scan_matcher/src/diagnostics_module.cpp
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/src/diagnostics_module.cpp]: fixed https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/2008b63ec46803a2f1216f26ba69bf5aa0e51624

File: localization/ndt_scan_matcher/src/map_update_module.cpp
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/src/map_update_module.cpp]: fixed https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/2008b63ec46803a2f1216f26ba69bf5aa0e51624

File: localization/ndt_scan_matcher/README.md
Comment (SakodaShintaro) [inline localization/ndt_scan_matcher/README.md:265]: "Transition condition to OK" is a bit confusing because the default condition of diagnostics is OK. 
I think this column may be delete.

File: localization/ndt_scan_matcher/README.md
Comment (SakodaShintaro) [inline localization/ndt_scan_matcher/README.md:326]: Strictly speaking, it is unclear what will happen if it is exactly equal to distance.

File: localization/ndt_scan_matcher/include/ndt_scan_matcher/ndt_scan_matcher_core.hpp
Comment (SakodaShintaro) [inline localization/ndt_scan_matcher/include/ndt_scan_matcher/ndt_scan_matcher_core.hpp:102]: If the function that performs the main processing of `service_ndt_align` is named `service_ndt_align_main`, should the function corresponding to `callback_initial_pose` be `callback_initial_pose_main`, and the function corresponding to `callback_sensor_points` should be `callback_sensor_points_main`?

**Review by SakodaShintaro — COMMENTED** [review]
I think there are generally no problems. I just added a few comments. I'll leave it up to you to deal with it.

File: localization/ndt_scan_matcher/include/ndt_scan_matcher/ndt_scan_matcher_core.hpp
Comment (YamatoAndo) [reply localization/ndt_scan_matcher/include/ndt_scan_matcher/ndt_scan_matcher_core.hpp]: fixed https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/fb5506937640da167b271a198a27b96f2477915f

File: localization/ndt_scan_matcher/README.md
Comment (SakodaShintaro) [inline localization/ndt_scan_matcher/README.md:281]: For some reason pre-commit is destroying the table :cry: 
https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/ce8adef9fa19d08d7b41b5f0aea967f43e028186

Comment (YamatoAndo) [reply localization/ndt_scan_matcher/README.md]: fixed https://github.com/autowarefoundation/autoware.universe/pull/5076/commits/3ccfa245e549be62cb3f89d98e6d7c13bfec1c4c

**Review by SakodaShintaro — APPROVED** [review]
Looks Good To Me

**Pull request merged**
